vortex-array/src/array/primitive/typed.rs

```diff
@@ -0,0 +1,70 @@
+//! TypedPrimitiveArray is a typed wrapper around `PrimitiveArray`.
+//!
+//! It provides ergonomics for cases where you can guarantee at compile time that a particular `
+//! PrimitiveArray` is of a certain type.
+//!
+//! Example:
+//!
+//! ```
+//! use vortex_array::array::TypedPrimitiveArray;
+//! use vortex_array::{IntoArray, IntoArrayVariant};
+//! use vortex_buffer::buffer;
+//!
+//! // Create a new array of values
+//! let values = buffer![1i32, 2, 3, 4];
+//! let values = values.into_array().into_primitive().unwrap();
+//! let typed_array = TypedPrimitiveArray::<i32>::try_from(values).unwrap();
+//!
+//! // Directly index the values like a normal slice.
+//! fn sum(values: impl AsRef<[i32]>) -> i32 {
+//!   values.as_ref().iter().sum()
+//! }
+//! assert_eq!(sum(&typed_array), 10i32);
+//! ```
+//!
+
+use std::ops::Deref;
+
+use vortex_dtype::NativePType;
+use vortex_error::{vortex_bail, VortexResult};
+
+use crate::array::PrimitiveArray;
+use crate::variants::PrimitiveArrayTrait;
+
+#[derive(Clone, Debug)]
+pub struct TypedPrimitiveArray<T> {
+    inner: PrimitiveArray,
+    _phantom: std::marker::PhantomData<T>,
+}
+
+impl<T: NativePType> TypedPrimitiveArray<T> {
```

Comment:
❤️

Reply:
Sadly I think I'm going to revert this piece b/c I didn't end up using it, but I do think we'd benefit from having something like this